File: javascript/countingSort.js
function countingSort(arr) {
  const max = Math.max(arr);
  const count = new Array(max + 1).fill(0);

  for(const num of arr) {
    count[num]++;
  }
  for(let i = 1; i <= max; i++) {
    count[i] += count[i - 1];
  }
  // build sorted array
  const output = new Array(arr.length);
  for(let i = arr.length - 1; i >= 0; i--) {
    output[count[arr[i]] - 1] = arr[i];
    count[arr[i]]--;
  }
  return output;
}

// example usage
const arr = [64, 25, 12, 22, 11];
const sortedArr = countingSort(arr);
console.log(sortedArr);

File: python/counting_sort.py
def counting_sort(arr):
  max_element = max(arr)

  # create count array to store count of each element
  count = [0] * (max_element + 1)

  # store count of each element in count array 
  for num in arr:
    count[num] += 1

  # modify count array to store actual position of element
  for i in range(1, max_element + 1):
    count[i] += count[i - 1]

  # build sorted array
  output = [0] * len(arr)
  for num in reversed(arr):
    output[count[num] - 1] = num
    count[num] -= 1

  return output

# example usage
arr = [64, 25, 12, 22, 11]
sorted_arr = counting_sort(arr)
print(sorted_arr)

